src/structs.py

```diff
@@ -649,6 +649,142 @@ def unset_option(self) -> None:
+        """Look up the option from the trajectory. Make sure consistent.
+        """
+        for i, act in enumerate(self.trajectory[1]):
+            if i == 0:
+                option = act.get_option()
+            else:
+                assert option == act.get_option()
+        self.set_option(option)
+        assert self.has_option()
+
+
+@dataclass(eq=False)
+class Partition:
+    """A partition stores a collection of segments that will ultimately be
+    covered by the same NSRT. For each segment, the partition also stores an
+    ObjToVarSub, under which the ParameterizedOption and effects for all
+    segments in the partition are equivalent.
+    """
+    members: List[Tuple[Segment, ObjToVarSub]]
+
+    def __iter__(self) -> Iterator[Tuple[Segment, ObjToVarSub]]:
+        return iter(self.members)
+
+    def __len__(self) -> int:
+        return len(self.members)
+
+    @cached_property
+    def _exemplar(self) -> Tuple[Segment, ObjToVarSub]:
+        assert len(self.members) > 0, "Partition is empty."
+        return self.members[0]
+
+    @cached_property
+    def add_effects(self) -> Set[LiftedAtom]:
+        """Get the lifted add effects for this partition.
+        """
+        seg, sub = self._exemplar
+        return {a.lift(sub) for a in seg.add_effects}
+
+    @cached_property
+    def delete_effects(self) -> Set[LiftedAtom]:
+        """Get the lifted delete effects for this partition.
+        """
+        seg, sub = self._exemplar
+        return {a.lift(sub) for a in seg.delete_effects}
+
+    @cached_property
+    def option_spec(self) -> Tuple[ParameterizedOption, List[Variable]]:
+        """Get the parameterized option and option vars for this partition.
+        """
+        seg, sub = self._exemplar
+        assert seg.has_option()
+        option = seg.get_option()
+        option_args = [sub[o] for o in option.objects]
+        return (option.parent, option_args)
+
+    def add(self, member: Tuple[Segment, ObjToVarSub]) -> None:
+        """Add a new member.
+        """
+        seg, sub = member
+        # Check for consistency.
+        if len(self.members) > 0:
+            # The effects should match.
+            lifted_add_effects = {a.lift(sub) for a in seg.add_effects}
+            lifted_delete_effects = {a.lift(sub) for a in seg.delete_effects}
+            assert lifted_add_effects == self.add_effects
+            assert lifted_delete_effects == self.delete_effects
+            if seg.has_option():
+                option = seg.get_option()
+                part_param_option, part_option_args = self.option_spec
+                assert option.parent == part_param_option
+                option_args = [sub[o] for o in option.objects]
+                assert option_args == part_option_args
+        # Add to members.
+        self.members.append(member)
+
+
 # Convenience higher-order types useful throughout the code
 ActionTrajectory = Tuple[List[State], List[Action]]
```

Comment:
I don't know if this is the best name for this. Why not StateActionTrajectory?

Reply:
@ashayathalye thanks, good suggestion. I'll do this in a separate PR later today.

Reply:
handled by #121